### DISCUSSION TITLE
Embedding in cljs/react/reagent app

### DISCUSSION THREAD

**Discussion by Ramblurr** [metasoarous/oz#166] [Q&A] [answered]
The docs say:

> Eventually we'll be adding options for hooking into the signal dataflow graphs within these visualizations so that interactions in a Vega/Vega-Lite visualization can be used to inform other Reagent components in your app.

Is this the current state of things? Is it possible to embed an interactivevega-lite visualizations into a cljs app such that it can react to the app and vice versa?

**Comment by metasoarous** [comment, marked as answer]
Hi @Ramblurr. Thanks for asking about this, and sorry for just not getting around to responding sooner.

The answer to this is a little complicated, so I'm moving this issue to a discussion, and will be forking off some issues from it.

It's been for quite a while possible for the vega visualization to respond to changes in react/reagent state (just use the `oz/vega` and `oz/vega-lite` reagent components as usual, basically). The downside of this is that it does require complete rerendering the visualization, which for larger datasets where just one or two data points (or other signal data) needs to change, could incur a more significant performance cost than we'd like. The [vega view api](https://vega.github.io/vega/docs/api/view/) does offer support for more granular state updates (in particular [`signal`](https://vega.github.io/vega/docs/api/view/#view_signal), [`setState`](https://vega.github.io/vega/docs/api/view/#view_setState), [`data`](https://vega.github.io/vega/docs/api/view/#view_data), [`change`](https://vega.github.io/vega/docs/api/view/#view_change), [`insert`](https://vega.github.io/vega/docs/api/view/#view_insert) and [`remove`](https://vega.github.io/vega/docs/api/view/#view_remove) methods), but till now, the view api hasn't been exposed via Oz.

On the flip side, in order to have the rest of a reagent app react to visualization events, there's no way around needed access to the view api (note [`addSignalListener`](https://vega.github.io/vega/docs/api/view/#view_addSignalListener), [`addEventListener`](https://vega.github.io/vega/docs/api/view/#view_addEventListener), [`addResizeListener`](https://vega.github.io/vega/docs/api/view/#view_addResizeListener), and [`addDataListener`](https://vega.github.io/vega/docs/api/view/#view_addDataListener)). 

This issue (#167) has now been resolved with 747187e, and has been released as `1.6.0-alpha35-SNAPSHOT`. It's not possible to pass a `:view-callback` function to the `vega` and `vega-lite` components which gets called with the vega `view` api instance once the visualization renders. For some functionality, this will require caching the `view` object somewhere for view later (in particular for triggering more granular single/data mutations), but is at least possible now.

I think there is still a lot of room for more convenient access to this kind of functionality. Some additional `:add-signal-listener` type options to the `vega` and `vega-lite` components would make things a bit easier. And on the mutation side, it would be amazing if:
* signals and data could be passed through to oz specifications as reactive atoms, and update that way
* install reactive atoms to capture signal changes originating from the visualization itself (brush selections, etc)
* possibly even support passing core.async channels for streaming data into a visualization (via [`insert`](https://vega.github.io/vega/docs/api/view/#view_insert))

I'll open up issues for these optimizations (#13 kind of exists, but misses the mark a bit), and update here. But in the mean time, please chime in here if you have any feedback about what any of this might look like, or if there are any questions about how things might work (or could be facilitated currently with the new `:view-callback` option).

Thanks again!
